### ISSUE TITLE
Doubt in GNN tutorial

### ISSUE THREAD

**Issue by anorak94** [phlippe/uvadlc_notebooks#6]
Tutorial 7 - 
Hey thanks for these wonderful notebooks. I was going through the GNN code especially the GAT section and just to be sure I understood everything correctly i replicated everything for your special case. 

        a_input = torch.cat([torch.index_select(input=node_feats_flat, index=edge_indices_row, 
               dim=0), torch.index_select(input=node_feats_flat, index=edge_indices_col, dim=0)
                ], dim=-1) 

        # Calculate attention MLP output (independent for each head)
        attn_logits = torch.einsum('bhc,hc->bh', a_input, self.a)
        attn_logits = self.leakyrelu(attn_logits)

In this line you stack the features of the nodes in each edge so say we have two nodes 
`i, j` we get a 2x2 matrix corresponding to them represented as 
`[[a, b] , [c, d]]` and we have the attention weights `[[w, x] , [y, z]]`. When we do the `einsum` operation we get a 2x1 matrix `[[aw+bx, cy+dz]]`. This is the first doubt shouldnt it be `[aw+bx+cy+dz]` according to the equation as for each `i, j` we have one value. If we have two heads then the `a matrix` should have shape `2x2*d` where `d=2` for our case

But going further down keeping the same calculations as above. After the attention probabilities. 

`node_feats = torch.einsum('bijh,bjhc->bihc', attn_probs, node_feats)` which is this line 

which can be expanded into where ap is the attention probabilites and feats is the node features after the linear projection

           for i in range(4):
               p1 = ap[i, :, 0]
               p2 = ap[i, :, 1]
      
               f1 = feats[:, 0, :].squeeze() ## dimension 1 
               f2 = feats[:, 1, :].squeeze() ## dimension 2 
               p1.shape , f1.shape, f2.shape
    
               print(torch.tensor([(torch.dot(p1, f1), torch.dot(p2, f2))]))

we see that the results is obtained by taking the two different probabilites from different heads and taking the dot product with two different dimensions of the feature matrix, but each head should operate on both the dimensions of the node features or atleast I hope it should. I check the output at each intermediate stage to be sure that it matches the results from the notebook you provide. Am i missing something

**Comment by phlippe** [comment]
Hi, thanks for raising the questions. Let me try to answer them below.

Regarding the first code fragment: you are right in your intuition, but the code in the notebook operates slightly different for efficiency reasons. Instead of calculating the attention logits for every possible pair of variables, we calculate the attention logit for every edge in all graphs. Hence, the tensor `a_input` has the shape `[number of edges, number of heads, 2*number of features per head]`. On this tensor, we then apply the einsum operation, and obtain a tensor attn_logits of shape `[number of edges, number of heads]`. Afterwards, we re-organize the tensor back into a matrix of shape `[number of batches, number of nodes, number of nodes]`, and take the softmax over the last dimension. If I understand your intuition correctly, you would have stuck to the matrix in the first place. This is more intuitive and simpler to implement, but quickly becomes inefficient when we have more zeros in the adjacency matrix than ones (i.e. any larger graph).

Regarding the second code fragment: you are right that each head only operates on its assigned features and not all channels. This is done on purpose and follows the proposal of the original paper ([Velickovic et al.](https://arxiv.org/abs/1710.10903)). If all heads operated on all features, we would have an N times larger feature output size, where N is the number of heads. The features can be combined between heads either in the next layer, or summed directly, as done for the last layer.

**Comment by anorak94** [comment]
Thanks for answering. So from what I understand you are saying we define an attention weight for each set of features so in our case we have two heads and the two heads operate on dimension 1 and 2 of the node features. So here `a_input is [number of edges, number of heads, 2*number of features per head] `  and for each edge we take the `2x2` matrix and multiply it elementwise with the `a` matrix 

`[[a, b] , [c, d]]` and we have the attention weights `[[w, x] , [y, z]]`. We do the elementwise multiplication which gives `[[aw, bx] , [cy, dz]]` and then sum along axis 1 which gives `[[aw+bx , cy+dz]]`. This is the attention logits for each feature and then we compute the probabilities for each head separately by softmaxing across dimension 2. And since each head operates on a different set of features we compute the final weighted node_feats by doing a weighted average of the respective features of node_feats with their respective attention_probs

**Comment by anorak94** [comment]
`prob_ij = exp(aw+cy)/ (exp(aw+cy) + sum other logits(i, k))` and say `prob_iu = exp(aw+mn)/ (exp(aw+mn) + sum other logits(i, k))` and then when we compute the weighted node features towards the end we replace the adjacency matrix with prob and this tells us how much we should weigh each neighbor of i in the neighborhood computation

**Comment by phlippe** [comment]
Yes, that is correct. Only correction: we sum along the last axis when doing the einsum operation over `a_input` and `a`, i.e. the `c` dimension. Hence, in your example, we would get `[aw+bx, cy+dz]` if I understand it correctly.

**Comment by anorak94** [comment]
Yes but then a is feature 1 in node 1 and b is feature 1 in node 2 and w and x are attention weights operating on dimension 1

**Comment by phlippe** [comment]
OK yes, then we mean the same thing here.

**Comment by anorak94** [comment]
Nice thanks a lot
